lib/src/connection_pool.dart

```diff
@@ -0,0 +1,51 @@
+part of mongo_dart;
+
+/// A function that produces an instance of [Db], whether synchronously or asynchronously.
+///
+/// This is used in the [ConnectionPool] class to connect to a database on-the-fly.
+typedef FutureOr<Db> DbFactory();
+
+/// A connection pool that limits the number of concurrent connections to a MongoDB server.
+///
+/// The connection pool lazily connects to the database; that is to say, it only opens as many
+/// connections as it needs to. If it is only ever called once, then it will only ever connect once.
+class ConnectionPool {
+  List<Db> _connections = [];
+  int _index = 0;
+  Pool _pool;
+
+  /// The maximum number of concurrent connections allowed.
+  final int concurrency;
+
+  /// A [DbFactory], a parameterless function that returns a [Db]. The function can be asynchronous if necessary.
+  final DbFactory dbFactory;
```

Comment:
None in particular. I can mark this as private, and then expand on the comment in the constructor to make it clearer.